otherlibs/stdune/src/loc0.ml

```diff
@@ -1,19 +1,133 @@
-include Lexbuf.Loc
+type t =
+  | No_loc
+  | In_file of string
+  | Lexbuf_loc of Lexbuf.Loc.t
+  | Same_file of
+      { pos_fname : string
+      ; start_lnum : int
+      ; start_bol : int
+      ; start_cnum : int
+      ; stop_lnum : int
+      ; stop_bol : int
+      ; stop_cnum : int
+      }
+  | Same_line of
+      { pos_fname : string
+      ; pos_lnum : int
+      ; pos_bol : int
+      ; start_cnum : int
+      ; stop_cnum : int
+      }
 
-let to_lexbuf_loc x = x
+module Position = Lexbuf.Position
+open Lexbuf.Loc
 
-let of_lexbuf_loc x = x
+let to_lexbuf_loc = function
+  | No_loc -> Lexbuf.Loc.none
+  | Lexbuf_loc loc -> loc
+  | In_file fname -> Lexbuf.Loc.in_file ~fname
+  | Same_file
+      { pos_fname
+      ; start_lnum
+      ; start_bol
+      ; start_cnum
+      ; stop_lnum
+      ; stop_bol
+      ; stop_cnum
+      } ->
+    let start =
+      { Lexing.pos_fname
+      ; pos_lnum = start_lnum
+      ; pos_bol = start_bol
+      ; pos_cnum = start_cnum
+      }
+    in
+    let stop =
+      { Lexing.pos_fname
+      ; pos_lnum = stop_lnum
+      ; pos_bol = stop_bol
+      ; pos_cnum = stop_cnum
+      }
+    in
+    { Lexbuf.Loc.start; stop }
+  | Same_line { pos_fname; pos_lnum; pos_bol; start_cnum; stop_cnum } ->
+    let start =
+      { Lexing.pos_fname; pos_lnum; pos_bol; pos_cnum = start_cnum }
+    in
+    let stop = { Lexing.pos_fname; pos_lnum; pos_bol; pos_cnum = stop_cnum } in
+    { Lexbuf.Loc.start; stop }
 
-let start t = t.start
+let in_file ~fname = In_file fname
 
-let stop t = t.stop
+let of_lexbuf_loc loc =
```

Comment:
Why do we need the `Lexbuf_loc` variant if we have conversions both ways?

Reply:
More generally, it seems odd that `Loc.t` depends on some `Lexbuf.Loc.t` instead of the other way around. If seems weird to special case `Lexbuf` in this way.

Reply:
> Why do we need the Lexbuf_loc variant if we have conversions both ways?

It's still needed for locations that span more than one file. It's essentially a fallback when we can't use an optimized representation. I do have an upcoming patch where the unoptimized version will be a bit more important.

> More generally, it seems odd that Loc.t depends on some Lexbuf.Loc.t instead of the other way around. If seems weird to special case Lexbuf in this way.

`Lexbuf.Loc.t` is a bit more simple I think. It's a stable representation we use for RPC and it comes straight from the lexer. So it makes sense that our specialized location type depends on it.

Reply:
I see. That makes sense though I think it would still be cleaner if we had a most general location type (based on two positions) isomorphic to what's currently called `Lexbuf.Loc.t`, with both `Lexbuf.Loc.t` and `Loc.t` using it.

This doesn't need to happen in this PR though.

Reply:
What would this general location type be called though? Right now we have

* Lexbuf.Loc.t for what the lexer produces
* Loc.t for what dune prefers to use

Presumably this general location type would have a top level name, so we'd need to pick a name that somehow distinguishes it.

Reply:
For example, the most general type could be `Loc.General.t`, and it could be used by both `Loc.t` and `Lexbuf.t`.